Add build documentation (#55)
rapidsai/build-planning: docs/docs/ci.md

# CI and Deployment

> All RAPIDS packages are built using the standard infrastructure leveraged by CI.

Reviewer: ```suggestion
All RAPIDS packages are built using a standard shared CI infrastructure.
```
We can make this less wordy

For wheels, separate images are created for build and testing.
While the build images contain all libraries required for building RAPIDS, the test image is slimmed down and is missing some of these libraries.
This allows us to test (at least to some extent) our manylinux compliance by verifying that our packages run even without those extra libraries installed.
In an ideal world the testing image would not even have the CTK installed, but that is not currently possible because RAPIDS has some dependencies (numba and cupy) that are not as self-contained and still rely on a system CUDA installation.
Until that changes, we cannot remove the CTK from wheel testing images.

### conda images

For conda, the images are different depending on whether they are for CUDA 11 or for CUDA 12+.
As described in [the conda section above](packaging.md#conda) the available packages differed between these versions.
Therefore, in RAPIDS conda CI images, for CUDA 11 images the CTK is installed on the system, while for CUDA 12 images it is not.
The installation of conda itself in our CI images is handled upstream in the miniforge-cuda images (see below).

> The [ci-imgs repository is automatically rebuilt whenever new changes are pushed to any of the following:

Reviewer 2: ```suggestion
The ci-imgs repository is automatically rebuilt whenever new changes are pushed to any of the following:
```

This looks like it was a mistake.

> The ci-imgs repository is automatically rebuild whenever new changes are pushed to any of the following:
> 
> - dependency-file-generator
> - miniforge-cuda
> - gha-tools

Reviewer 2: ```suggestion
The [ci-imgs](https://github.com/rapidsai/ci-imgs) repository is automatically rebuild whenever new changes are pushed to any of the following:

- [dependency-file-generator](https://github.com/rapidsai/dependency-file-generator)
- [miniforge-cuda](https://github.com/rapidsai/miniforge-cuda)
- [gha-tools](https://github.com/rapidsai/gha-tools)
```

Could you please make all of these links?

For example, I think this is the first time in this documentation that `gha-tools` has been linked. Might not be immediately clear to someone new that this is a repo in the `rapidsai` GitHub org.

Author: Ah yeah so this is a mixed bag:
- ci-imgs is linked above. This is a subsection of the ci-imgs section, so no need to relink.
- dependency-file-generator is covered on a separate page that is technically sequenced before, but there is obviously no guarantee that someone will read this in sequence. Not linking here is an artifact of when I originally had one big doc before I split it up. I'll link to the dfg section in the docs instead.
- miniforge-cuda is just below this. I'll just add a "see below"
- gha-tools is in the previous section. If you think that's too far away and it might be missed, we could link back to the section, but I feel like it's good enough that it's in the same page since someone just doing Ctrl + F on the page will find it. WDYT?

Reviewer 2: Yep all works for me! I think it's right to not assume a specific sequence of reading these docs.

The linking and parentheticals you added in https://github.com/rapidsai/build-planning/pull/55/commits/f92209a043c7117005c054f75edcd22eb94558ce address what I was going for with this comment, I'm good with this. Thank you!

This ensures that the images always have the latest version of everything required by our CI build scripts.

#### miniforge-cuda

Our conda CI images are based on the [miniforge-cuda images](https://github.com/rapidsai/miniforge-cuda), which we also build.
These images are based off of [`nvidia/cuda`](https://hub.docker.com/r/nvidia/cuda) Docker images, like the wheel images.
They are a minimal installation of [`miniforge`](https://github.com/conda-forge/miniforge) on top of the CUDA images to enable conda in systems with the CTK installed.

## Workflows

### Runners

> RAPIDS runs Github actions on [a set of self-hosted runners at NVIDIA](https://docs.gha-runners.nvidia.com/).

Reviewer: ```suggestion
RAPIDS runs Github actions on a mixture of public and private runners[self-hosted runners at NVIDIA](https://docs.gha-runners.nvidia.com/).
```
We have public s3 runners still ( IIRC for arm )

Author: Yeah we definitely still run some jobs on public runners.
